index.src.html

```diff
@@ -86,42 +86,57 @@
+perspective of web authors. Implementers may choose to use more or less threads
+as long the exernally visible behaviors of blocking and sequencing are
+maintained as follows.
+
+Control Thread and Codec Thread {#control-thread-and-codec-thread}
+------------------------------------------------------------------
 
-The steps of a control message may depend on <dfn>injected state</dfn>, supplied
-by the method that enqueued the message.
+All steps in this specificaiton will run on either a [=control thread=] or
+a [=codec thread=].
 
-<dfn lt="Run the control message processing loop">Running the control message
-    processing loop</dfn> means performing these steps.
-1. While the control message queue is not <a>empty</a>
-    1. Let |front message| be the next <a>dequeued</a> <a>control message</a>
-    2. If |front message| cannot be performed now, return.
+The <dfn>control thread</dfn> is the thread from which authors will construct
+and [=codec=] and invoke its methods. Invoking a codec's methods will typically
+result in the creation of [=control messages=] which are later executed on the
+[=codec thread=]. Each [=JavaScript realm=] has a separate control thread.
 
-        The User Agent must decide when further processing is blocked because of
-            ongoing work as an implementation detail (e.g. the underlying
-            decoder cannot accept more requests yet). The UA must restart the
-            processing loop when the blockage is resolved.
+The <dfn>codec thread</dfn> is the thread from which a [=codec=] will
+[=dequeue=] [=control messages=] and execute their steps. Each [=codec=]
+instance has a separate codec thread.
 
-        NOTE: a blocked processing loop is visible to authors via the
-            `decodeQueueSize` and `encodeQueueSize` attributes.
+The [=control thread=] uses a traditional event loop, as described in [[!HTML]].
 
-    3. Dequeue |front message| from the <a>control message queue</a>.
-    4. Run the |front message| <a>control message steps</a>.
+The [=codec thread=] uses a specialized [=codec processing loop=].
+
+Communication from the [=control thread=] to the [=codec thread=] is done using
+[=control message=] passing. Communication in the other direction is done using
+regular event loop tasks.
+
+Each [=codec=] instance has a single <dfn>control message queue</dfn> that is
+a [=queue=] of <dfn>control messages</dfn>.
+
+<dfn lt="Enqueues a control message|Queue a control message">Queuing a control
+message</dfn> means [=enqueing=] the message to a [=codec=]’s [=control
+message queue=]. Invoking codec methods will often queue a control message
+to schedule work.
+
+<dfn lt="running a control message|control message steps">Running a control
+message</dfn> means performing a sequence of steps specified by the method
+that enqueued the message. The steps of a control message may depend on
+<dfn>injected state</dfn>, supplied by the method that enqueued the message.
```

Comment:
This kind of ties into garbage collection and references, or are you thinking purely about explicit name binding to make it extra clear what is what ?

It seems to me that this is too specific, and we can just shepherd objects as we go along maybe? To take an example from the Web Audio API, https://webaudio.github.io/web-audio-api/#dom-audiocontext-suspend, we have a promise created on the control thread, then we do some stuff on the audio thread, then we fire a task back to the control thread and then we resolve the promise, and it's not really ambiguous _to me_ what happens (but I'm biased, I wrote it 😃).

Reply:
Agree, that sounds cleaner. Filed https://github.com/WICG/web-codecs/issues/136